File: CANON_SDK/EDSDK_v13.19.0_Macintosh/Sample/Swift/CameraControl/CameraControl/FileCounterCommand.swift
/******************************************************************************
 *                                                                             *
 *   PROJECT : Eos Digital camera Software Development Kit EDSDK               *
 *                                                                             *
 *   Description: This is the Sample code to show the usage of EDSDK.          *
 *                                                                             *
 *                                                                             *
 *******************************************************************************
 *                                                                             *
 *   Written and developed by Canon Inc.                                       *
 *   Copyright Canon Inc. 2018 All Rights Reserved                             *
 *                                                                             *
 *******************************************************************************/

import Foundation

class FileCounterCommand : Command{
    
    //Count the files in the specified volume.
    func countImages(_ camera : EdsCameraRef, volumeIndex : EdsUInt32, targetFileName : inout String, outCount : inout Int, outItem : inout EdsBaseRef?, outImageItems: inout Array<EdsDirectoryItemRef>)->EdsError{
        
        var volumeCount = 0 as EdsUInt32
        var error = EdsGetChildCount(camera, &volumeCount)
        
        if (error != EdsError(EDS_ERR_OK)){
            return error
        }
        if(volumeCount <= volumeIndex){
            return EdsError(EDS_ERR_DIR_NOT_FOUND)
        }
        
        var volume : EdsBaseRef? = nil
        error = EdsGetChildAtIndex(camera, EdsInt32(volumeIndex), &volume)
        if (error != EdsError(EDS_ERR_OK)){
            return error
        }
        
        var itemCount = 0 as EdsUInt32
        error = EdsGetChildCount(volume, &itemCount)
        if (error != EdsError(EDS_ERR_OK)){
            return error
        }
        
        for itemIndex in 0..<itemCount{
            var dirItem : EdsBaseRef? = nil
            error = EdsGetChildAtIndex(volume, EdsInt32(itemIndex), &dirItem)
            if (error != EdsError(EDS_ERR_OK)){
                continue
            }
            
            var dirItemInfo = EdsDirectoryItemInfo()
            error = EdsGetDirectoryItemInfo(dirItem, &dirItemInfo)
            if (error != EdsError(EDS_ERR_OK)){
                return error
            }
            
            let itemName = withUnsafePointer(to: &dirItemInfo.szFileName) {
                String(cString: UnsafeRawPointer($0).assumingMemoryBound(to: CChar.self))
            }
            
            if(itemName == targetFileName && dirItemInfo.isFolder == 1){
                outItem = dirItem!
                break
            }
            
            if(dirItem != nil){
                EdsRelease(dirItem)
            }
        }
        
        var dirCount = 0 as EdsUInt32
        error = EdsGetChildCount(outItem, &dirCount)
        
        for dirIndex in 0..<dirCount{
            var dirItem : EdsDirectoryItemRef? = nil
            error = EdsGetChildAtIndex(outItem, EdsInt32(dirIndex), &dirItem)
            if (error != EdsError(EDS_ERR_OK)){
                continue
            }
            
            var fileCount = 0 as EdsUInt32
            error = EdsGetChildCount(dirItem, &fileCount)
            if (error != EdsError(EDS_ERR_OK)){
                return error
            }
            outCount += Int(fileCount)
            
            for fileIndex in 0..<fileCount{
                var item : EdsDirectoryItemRef? = nil
                error = EdsGetChildAtIndex(dirItem, EdsInt32(fileIndex), &item)
                if (error != EdsError(EDS_ERR_OK)){
                    continue
                }
                outImageItems.append(item!)
            }
        }
        return error
    }
    
    func changeTargetFileName(targetFileName : inout String){
        if (targetFileName == "DCIM"){
            targetFileName = "XFVC"
        }
        else if (targetFileName == "XFVC"){
            targetFileName = ""
        }
    }
}

File: src/CANON_SDK/EDSDK_v13.19.0_Windows/sample/CSharp/CameraControl/CameraControl/Command/FileCounterCommand.cs
/******************************************************************************
*                                                                             *
*   PROJECT : Eos Digital camera Software Development Kit EDSDK               *
*                                                                             *
*   Description: This is the Sample code to show the usage of EDSDK.          *
*                                                                             *
*                                                                             *
*******************************************************************************
*                                                                             *
*   Written and developed by Canon Inc.                                       *
*   Copyright Canon Inc. 2018 All Rights Reserved                             *
*                                                                             *
*******************************************************************************/

using System;
using System.CodeDom;
using System.Collections.Generic;
using System.Security.Cryptography;

namespace CameraControl
{
    class FileCounterCommand : Command
    {
        public FileCounterCommand(ref CameraModel model, ref IntPtr volume) : base(ref model) { _volume = volume; }

        public struct FileNumber
        {
            public IntPtr DirItem;
            public int dirnum;
            public int[] filenum;

            public FileNumber(int p1)
            {
                DirItem = IntPtr.Zero;
                dirnum = p1;
                filenum = new int[dirnum];
            }
        }

        private IntPtr _volume;

        public uint CountDirectory(IntPtr camera, ref IntPtr directoryItem, ref string targetFolderName, out int directory_count)
        {
            uint err = EDSDKLib.EDSDK.EDS_ERR_OK;
            int item_count = 0;
            directory_count = 0;

            // Get any folder
            IntPtr dirItem;
            EDSDKLib.EDSDK.EdsDirectoryItemInfo dirItemInfo;
            dirItemInfo.szFileName = "";
            dirItemInfo.Size = 0;

            err = EDSDKLib.EDSDK.EdsGetChildCount(_volume, out item_count);
            if (err != EDSDKLib.EDSDK.EDS_ERR_OK)
            {
                return err;
            }

            for (int i = 0; i < item_count; ++i)
            {
                // Get the ith item under the specifed volume
                err = EDSDKLib.EDSDK.EdsGetChildAtIndex(_volume, i, out dirItem);
                if (err != EDSDKLib.EDSDK.EDS_ERR_OK)
                {
                    continue;
                }

                // Get retrieved item information
                err = EDSDKLib.EDSDK.EdsGetDirectoryItemInfo(dirItem, out dirItemInfo);
                if (err != EDSDKLib.EDSDK.EDS_ERR_OK)
                {
                    return err;
                }

                // Indicates whether or not the retrieved item is a DCIM or XFVC folder.
                if (dirItemInfo.szFileName == targetFolderName && dirItemInfo.isFolder == 1)
                {
                    directoryItem = dirItem;
                    break;
                }

                // Release retrieved item
                if (dirItem != IntPtr.Zero)
                {
                    EDSDKLib.EDSDK.EdsRelease(dirItem);
                }
            }

            // Get number of directory
            return err = EDSDKLib.EDSDK.EdsGetChildCount(directoryItem, out directory_count);

        }

        // Due to the addition of XFVC specifications, a function is provided to change the target folder.
        public void ChangeTargetFolderName(ref string targetFolderName)
        {
            if (targetFolderName == "DCIM")
            {
                targetFolderName = "XFVC";
            }
            else if (targetFolderName == "XFVC")
            {
                targetFolderName = "";
            }
        }

        public uint CountImages(IntPtr camera, ref IntPtr directoryItem, ref int directory_count, ref int fileCount, ref FileNumber fileNumber, ref List<IntPtr> imageItems)
        {
            uint err = EDSDKLib.EDSDK.EDS_ERR_OK;

            // Get the number of camera volumes
            fileCount = 0;

            // Get retrieved item information

            for (int i = 0; i < directory_count; ++i)
            {
                int count = 0;
                err = CountImagesByDirectory(ref directoryItem, i, ref count, ref imageItems);
                if (err != EDSDKLib.EDSDK.EDS_ERR_OK)
                {
                    return err;
                }
                fileCount += count;
                fileNumber.filenum[i] = count;
            }
            return EDSDKLib.EDSDK.EDS_ERR_OK;
        }

        private uint CountImagesByDirectory(ref IntPtr directoryItem, int directoryNo, ref int image_count, ref List<IntPtr> imageItems)
        {
            int item_count = 0;

            IntPtr directoryfiles;
            IntPtr fileitem;
            EDSDKLib.EDSDK.EdsDirectoryItemInfo dirItemInfo;

            uint err = EDSDKLib.EDSDK.EdsGetChildAtIndex(directoryItem, directoryNo, out directoryfiles);
            if (err != EDSDKLib.EDSDK.EDS_ERR_OK)
            {
                return err;
            }

            // Get retrieved item information
            // Get item name
            err = EDSDKLib.EDSDK.EdsGetDirectoryItemInfo(directoryfiles, out dirItemInfo);
            if (err != EDSDKLib.EDSDK.EDS_ERR_OK)
            {
                return err;
            }

            int index = 0, filecount = 0;
            err = EDSDKLib.EDSDK.EdsGetChildCount(directoryfiles, out item_count);
            if (err != EDSDKLib.EDSDK.EDS_ERR_OK)
            {
                return err;
            }
            for (index = 0; index < item_count; ++index)
            {
                err = EDSDKLib.EDSDK.EdsGetChildAtIndex(directoryfiles, index, out fileitem);
                if (err != EDSDKLib.EDSDK.EDS_ERR_OK)
                {
                    return err;
                }

                // Get retrieved item information
                err = EDSDKLib.EDSDK.EdsGetDirectoryItemInfo(fileitem, out dirItemInfo);
                if (err != EDSDKLib.EDSDK.EDS_ERR_OK)
                {
                    return err;
                }
                if (dirItemInfo.isFolder == 0)
                {
                    imageItems.Add(fileitem);
                    filecount += 1;
                }

            }
            image_count = filecount;

            return EDSDKLib.EDSDK.EDS_ERR_OK;
        }


    }

}
